composeApp/src/commonMain/kotlin/org/ooni/probe/domain/FetchGeoIpDbUpdates.kt

```diff
@@ -0,0 +1,107 @@
+    }
+
+    suspend operator fun invoke(): Result<Path?, Engine.MkException> =
+        getLatestEngineVersion()
+            .onSuccess { version ->
+                val (isLatest, latestVersion) = isGeoIpDbLatest(version)
+                if (isLatest) {
+                    return Success(null)
+                } else {
+                    val url = buildGeoIpDbUrl(latestVersion)
+                    val target = "$cacheDir/$latestVersion.mmdb"
+
+                    downloadFile(url, target)
+                        .onSuccess { downloadedPath ->
+                            preferencesRepository.setValueByKey(
+                                SettingsKey.MMDB_VERSION,
+                                latestVersion,
+                            )
+                            preferencesRepository.setValueByKey(
+                                SettingsKey.MMDB_LAST_CHECK,
+                                Clock.System.now().toEpochMilliseconds(),
+                            )
+                            return Success(downloadedPath)
+                        }.onFailure { downloadError ->
+                            return Failure(Engine.MkException(downloadError))
+                        }
+                }
+            }.onFailure { versionError ->
+                return Failure(versionError)
+            }.let { Failure(Engine.MkException(Throwable("Unexpected state"))) }
+
+    /**
+     * Compare latest and current version integers and return pair of latest state and actual version number
+     * @return Pair<Boolean, String> where the first element is true if the DB is the latest,
+     * the second is the current version and the third is the latest version.
+     */
+    private suspend fun isGeoIpDbLatest(latestVersion: String): Pair<Boolean, String> {
+        val currentGeoIpDbVersion: String =
+            (
+                preferencesRepository.getValueByKey(SettingsKey.MMDB_VERSION).first()
+                    ?: GEOIP_DB_VERSION_DEFAULT
+            ) as String
+
+        return Pair(
+            normalize(currentGeoIpDbVersion) >= normalize(latestVersion),
+            latestVersion,
+        )
+    }
+
+    private suspend fun getLatestEngineVersion(): Result<String, MkException> {
+        val url = "https://api.github.com/repos/${GEOIP_DB_REPO}/releases/latest"
+
+        return engineHttpDo("GET", url, TaskOrigin.OoniRun).map { payload ->
+            payload?.let {
+                try {
+                    json.decodeFromString(GhRelease.serializer(), payload).tag
+                } catch (e: SerializationException) {
+                    Logger.e(e) { "Failed to decode release info" }
+                    null
+                } catch (e: IllegalArgumentException) {
+                    Logger.e(e) { "Failed to decode  release info" }
+                    null
+                }
+            } ?: throw MkException(Throwable("Failed to fetch latest version"))
```

Comment:
This will not return a Failure<MkException>, but will make `getLatestEngineVersion()` throw an exception instead, which is not handled.